drv/vsc7448/src/spi.rs

```diff
@@ -0,0 +1,276 @@
+    /// if you want to modify it, then use [Self::modify] instead.
+    pub fn write<T>(
+        &self,
+        reg: RegisterAddress<T>,
+        value: T,
+    ) -> Result<(), VscError>
+    where
+        u32: From<T>,
+    {
+        assert!(reg.addr >= 0x71000000);
+        assert!(reg.addr <= 0x72000000);
+
+        let addr = (reg.addr & 0x00FFFFFF) >> 2;
+        let value: u32 = value.into();
+        let data: [u8; 7] = [
+            0x80 | ((addr >> 16) & 0xFF) as u8,
+            ((addr >> 8) & 0xFF) as u8,
+            (addr & 0xFF) as u8,
+            ((value >> 24) & 0xFF) as u8,
+            ((value >> 16) & 0xFF) as u8,
+            ((value >> 8) & 0xFF) as u8,
+            (value & 0xFF) as u8,
+        ];
+
+        ringbuf_entry_masked!(
+            DEBUG_TRACE_SPI,
+            Trace::Write {
+                addr: reg.addr,
+                value: value.into()
+            }
+        );
+        self.0.write(&data[..])?;
+        Ok(())
+    }
+
+    /// Writes to a port mask, which is assumed to be a pair of adjacent
+    /// registers representing all 53 ports.
+    pub fn write_port_mask<T>(
+        &self,
+        mut reg: RegisterAddress<T>,
+        value: u64,
+    ) -> Result<(), VscError>
+    where
+        T: From<u32>,
+        u32: From<T>,
+    {
+        self.write(reg, ((value & 0xFFFFFFFF) as u32).into())?;
+        reg.addr += 4; // Good luck!
+        self.write(reg, (((value >> 32) as u32) & 0x1FFFFF).into())
+    }
+
+    /// Performs a write operation on the given register, where the value is
+    /// calculated by calling f(0).  This is helpful as a way to reduce manual
+    /// type information.
+    pub fn write_with<T, F>(
+        &self,
+        reg: RegisterAddress<T>,
+        f: F,
+    ) -> Result<(), VscError>
+    where
+        T: From<u32>,
+        u32: From<T>,
+        F: Fn(&mut T),
+    {
+        let mut data = 0.into();
+        f(&mut data);
+        self.write(reg, data)
+    }
+
+    /// Performs a read-modify-write operation on a VSC7448 register
+    pub fn modify<T, F>(
+        &self,
+        reg: RegisterAddress<T>,
+        f: F,
+    ) -> Result<(), VscError>
+    where
+        T: From<u32>,
+        u32: From<T>,
+        F: Fn(&mut T),
+    {
+        let mut data = self.read(reg)?;
+        f(&mut data);
+        self.write(reg, data)
+    }
+
+    /// Reads from a specific SERDES6G instance, which is done by writing its
+    /// value (as a bitmask) to a particular register with a read flag set,
+    /// then waiting for the flag to autoclear.
+    pub fn serdes6g_read(&self, instance: u32) -> Result<(), VscError> {
+        let mut reg: vsc7448_pac::hsio::mcb_serdes6g_cfg::MCB_SERDES6G_ADDR_CFG =
+            0.into();
+        reg.set_serdes6g_rd_one_shot(1);
+        reg.set_serdes6g_addr(1 << instance);
+        let addr = Vsc7448::HSIO().MCB_SERDES6G_CFG().MCB_SERDES6G_ADDR_CFG();
+        self.write(addr, reg)?;
+        for _ in 0..32 {
```

Comment:
This fixed iteration count seems suspicious. Should it be expressed in terms of wall time? Either way, a comment would be good.

Reply:
(same in other cases in this file)

Reply:
Added some documentation about this and moved it to a constant.  There's a pending `TODO` to see whether this actually takes more than one iteration...